components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -226,20 +270,38 @@ auto validate_node_id_value_pairs(
                 return std::errc::operation_not_permitted;
             }
 
-            // We checked that the node isn't the root above, so we can query the underlying ID
-            // safely without a repeated check.
-            auto const parent_node_id{node.get_parent_id_unsafe()};
-            auto const key_name{node.get_key_name()};
-            if (parent_node_id_to_key_names.contains(parent_node_id)) {
-                auto const [it, new_key_inserted]{
-                        parent_node_id_to_key_names.at(parent_node_id).emplace(key_name)
-                };
-                if (false == new_key_inserted) {
-                    // The key is duplicated under the same parent
+            if (false
+                == check_key_uniqueness_among_sibling_nodes(node, parent_node_id_to_key_names))
+            {
+                return std::errc::protocol_not_supported;
+            }
+
+            // Iteratively check if there's any key duplication in ancestors until:
+            // 1. The ancestor is already checked. We only need to check an ancestor node once. If
+            //    there are key duplications among its sibling, it will be caught when the sibling
+            //    is first checked. The order of which sibling gets checked first doesn't affect the
+            //    results.
+            // 2. Reached the root node.
```

Comment:
```suggestion
            // Iteratively check if there's any key duplication in the node's ancestors until:
            // 1. The ancestor has already been checked. We only need to check an ancestor node
            //    once since if there are key duplications among its siblings, it would've been
            //    caught when the sibling was first checked (the order in which siblings get checked
            //    doesn't affect the results).
            // 2. We reach the root node.
```